### PR TITLE
[main][feat] Add smart treasury keeper

### PR DESCRIPTION
**Pull request by m100alpaca**

### CODE REVIEW

File: contracts/8.12/SmartTreasuryDistributeKeeper.sol [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  function performUpkeep(
+    bytes calldata /* _performData */
+  ) external onlyIntervalPassed {
+    // Effect
+    lastTimestamp = block.timestamp;
+
+    // Interaction
+
+    // 1. Withdraw from money market to smart treasury
+    moneyMarket.withdrawProtocolReserves(withdrawProtocolReserveParams);
```
Comment (alpanaca) [inline contracts/8.12/SmartTreasuryDistributeKeeper.sol:60]: Shouldn't this be build every time? how would you know the amount

Comment (m100alpaca) [reply contracts/8.12/SmartTreasuryDistributeKeeper.sol:60]: Should we change from specific amount to withdraw all of it to treasury?

File: contracts/8.12/interfaces/IAdminFacet.sol [hunk 2]
```diff
@@ -0,0 +1,140 @@
+// SPDX-License-Identifier: BUSL
+pragma solidity 0.8.12;
+
+import { LibConstant } from "../libs/LibConstant.sol";
+
+// ---- Interfaces ---- //
+import { IFeeModel } from "./IFeeModel.sol";
+
+interface IAdminFacet {
```
Comment (alpanaca) [inline contracts/8.12/interfaces/IAdminFacet.sol:9]: Just merge this to IMoneyMarket. Get rid of everything else but the functions and struct that need to be used

File: contracts/8.12/interfaces/IFeeModel.sol [hunk 3]
```diff
@@ -0,0 +1,9 @@
+// SPDX-License-Identifier: BUSL
+pragma solidity 0.8.12;
+
+interface IFeeModel {
```
Comment (alpanaca) [inline contracts/8.12/interfaces/IFeeModel.sol:4]: This file can be removed once you removed all unrelated functions

File: contracts/8.12/libs/LibConstant.sol [hunk 4]
```diff
@@ -0,0 +1,22 @@
+// SPDX-License-Identifier: BUSL
+pragma solidity 0.8.12;
+
+library LibConstant {
```
Comment (alpanaca) [inline contracts/8.12/libs/LibConstant.sol:4]: This can also be removed

File: contracts/8.12/SmartTreasuryDistributeKeeper.sol (hunk 1 above)
Comment (m100alpaca) [reply contracts/8.12/SmartTreasuryDistributeKeeper.sol]: solved

File: contracts/8.12/SmartTreasuryDistributeKeeper.sol [hunk 5]
```diff
@@ -24,7 +24,7 @@ contract SmartTreasuryDistributeKeeper is IntervalKeepers {
 
   address[] public distributedTokens;
 
-  IMoneyMarket public moneyMarket;
+  IMoneyMarket public immutable moneyMarket;
   ISmartTreasury public smartTreasury;
```
Comment (wow-alpaca) [inline contracts/8.12/SmartTreasuryDistributeKeeper.sol:28]: if both are not modifiable, apply `immutable` on both state then

File: scripts/configs/exec/smart-treasury-keeper/set-withdraw-and-distributed-tokens.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import { ethers } from "hardhat";
+import { SmartTreasuryDistributeKeeper__factory } from "../../../../typechain";
+import { BigNumberish, Bytes } from "ethers";
+
+async function main() {
+  const deployer = (await ethers.getSigners())[0];
+  // TODO: pending smart treasury address
```
Comment (jr-alpaca) [inline scripts/configs/exec/smart-treasury-keeper/set-withdraw-and-distributed-tokens.ts:7]: no need to put todo. make an input constant 
const smartTreasuryAddress = "0x0000000000000000000000000000000000000000"
```suggestion
```

File: scripts/configs/exec/smart-treasury-keeper/set-withdraw-and-distributed-tokens.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  const smartTreasuryDistributeKeeper =
+    SmartTreasuryDistributeKeeper__factory.connect("0x", deployer);
+
+  // TODO: pending distributed tokens
+  const WBNB = "0x";
+  const DISTRIBUTED_TOKENS: Array<string> = [WBNB];
```
Comment (jr-alpaca) [inline scripts/configs/exec/smart-treasury-keeper/set-withdraw-and-distributed-tokens.ts:13]:
```suggestion
  const DISTRIBUTED_TOKENS: Array<string> = ["0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c"];
```

**Review by wow-alpaca — APPROVED** [review]

**Pull request merged**